text/0000-additional-app-metadata.md

```diff
@@ -0,0 +1,130 @@
+- Health check information can be set with [`HEALTHCHECK`](https://docs.docker.com/reference/dockerfile/#healthcheck)
+
+# Motivation
+[motivation]: #motivation
+
+- It is not presently possible to set this metadata using buildpacks.
+- Deploying with Docker or other container orchestrators, the system can use this information to more easily run images.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes that we add additional extension points so that it is possible to expose port and health check information in the application container images that are generated.
+
+As a buildpack author, I will be able to include port and health check information as part of the process type that are defined by a buildpack. 
+
+For each process type defined, I can add a list of ports. A port is defined as the port number and an optional protocol. If the protocol is not set, then `TCP` is assumed. To expose both `TCP` and `UDP`, then the port number should be included twice. Once with with each protocol. Valid values for the protocol are `TCP` and `UDP`. This is the same behavior as Docker's `EXPOSE` functionality, which is intentional, as it should enable buildpack metadata to map in the same way as `Dockerfile` metadata.
+
+For each process type defined, I can add health check information. The following health check properties will be available: interval, timeout, start period, start interval, retries, and command. If no health check information is provided, then nothing will be added to the image.
+
+Because the image can only have a single set of ports and health check information, the lifecycle will include the information for the default process type when it exports the application image.
+
+As an end user, if I want to have different ports/health check information exposed on the application image then the user will set the `--default-process` flag to `pack build` and change the default process type.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+As a buildpack author, additional information can be written into the [`launch.toml` file](https://github.com/buildpacks/spec/blob/main/buildpack.md#launchtoml-toml). This will include an array of ports and the health check information as described above.
+
+Example:
+
+```
+[[processes]]
+type = "web"
+command = ["java"]
+args = ["-jar", "app.jar"]
+default = true
+working-dir = "/workspace"
+
+[[processes.ports]]
+port = 8080
+protocol = "TCP"
+
+[[processes.ports]]
+port = 8081  # protocol is optional, defaults to "TCP"
+
+[processes.health-check]
+interval = "20s"
+timeout = "3s"
+start-period = "1m"
+start-interval = "30s"
+retries = 3
+use-shell = true
+command = "thc"
+```
+
+This information is written into the `Config` section of the image metadata.
+
+```
+{
+    ...
+    "Config": {
+        ...
+        "ExposedPorts": {
+            "8080/tcp": {},
+            "9000/udp": {}
+        },
+        "Healthcheck": {
```

Comment:
I am curious about how the OCI image spec deals with [Healthchecj](https://github.com/opencontainers/image-spec/blob/v1.1.0/config.md#properties)

> Healthcheck object, OPTIONAL
> This property is reserved for use, to [maintain compatibility](https://github.com/opencontainers/image-> spec/blob/v1.1.0/media-types.md#compatibility-matrix).

Does **reserved for use** mean, we can use it, just to be compatible with Docker format, but it could change in the future to be in some other place in OCI image spec?

Reply:
To me, it sounds like they are reserving it to prevent conflicts. It's not part of the OCI spec, but the OCI spec authors seem concerned enough with compatibility that they don't want to overlap with certain popular things outside the spec.

My read is that this is OK to use if the intent is for compatibility with Docker. If we were introducing our own format for the health check data or something that could conflict with Docker, then it would not be OK.